Feature/class

afaikafk opened the pull request:
This PR essentially rewrites the greentasks library completely. The new design of it will be compatible with the previous API, but it will also allow the implementation of class based tasks, which can dynamically calculate their reschedule time (in case of periodic tasks), allow callers to specifiy callback and errback functions, and the tasks itself get packaged into a wrapper `PackagedTast` instance, which has a future object on it that can be used to obtain the return value or the exception that was raised during the asynchronous execution of the task.

afaikafk committed:
Add missing docstrings

afaikafk committed:
Rewrite library to support class based tasks along with simple callable
objects. Class based tasks provide the following opportunities:
- Dynamic delay for periodic tasks
- Periodic tasks can decide when to stop their re-scheduling once and
  for all
- Support callbacks and errbacks
- Support futures
- Provide better tracking of task status

afaikafk committed:
Remove unused import

afaikafk committed:
Make future type a class attribute on PackagedTask

afaikafk committed:
Make unique id generator return numbers instead of strings so it can be used by the hash method

afaikafk committed:
Implement __eq__ magic method as well

afaikafk committed:
Implement remaining unittests

afaikafk committed:
Perform task class conversion within PackagedTask

afaikafk committed:
Clarify the fate of delay and periodic parameters in the presence of task classes

afaikafk committed:
-Redesign task flow so that calculated delay from get_delay is carried on
between instances of tasks.
-Add optional name attribute to tasks
-Add ability to specify a different start delay, making it possible for
periodic tasks to run immediately when scheduled and then following the
calculated delay

afaikafk committed:
Fix storage of previous delay so it's not written into the class attribute, but stored on the packaged_task instance and passed as argument to get_delay of the next instance

afaikafk committed:
Add package level imports to make imports a joy

afaikafk committed:
Update setup module, fixing license info and bumping the version along the way

afaikafk committed:
Minor docstring and comment corrections

afaikafk committed:
Add API docs

afaikafk committed:
Fix license file contents

afaikafk committed:
Add basic usage instructions to readme

afaikafk committed:
Move autogenerated task name prefix out of method into class attribute

afaikafk committed:
Fix order of name and id in logging call

goodbadwolf commented on greentasks/scheduler.py:
```diff
@@ -1,94 +1,154 @@
-        return self.NOT_FOUND
+    def _execute(self, packaged_task):
+        """
+        Delegate execution of ``packaged_task`` to py:meth:`~PackageTask.run`,
```
Typo: Package_d_Task

goodbadwolf commented on greentasks/scheduler.py:
```diff
@@ -1,94 +1,154 @@
+    def _periodic(self, packaged_task):
+        """
+        Execute a periodic task through py:meth:`~TaskScheduler._execute` and
+        reschedule it automatically for the specified ``delay``.
```
`delay` is not a specified parameter

afaikafk replied:
Thanks, fixed

On Mon, Jun 13, 2016 at 6:47 PM, Manish P Mathai notifications@github.com
wrote:

> In greentasks/scheduler.py
> https://github.com/Outernet-Project/greentasks/pull/1#discussion_r66826394
> :
> 
> > -    def _execute(self, task_id, fn, args, kwargs):
> > -        self.current_task = task_id
> > -    def _periodic(self, packaged_task):
> > -        """
> > -        Execute a periodic task through py:meth:`~TaskScheduler._execute` and
> > -        reschedule it automatically for the specified `delay`.
> 
> delay is not a specified parameter
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/Outernet-Project/greentasks/pull/1/files/66f6c9e0d2f4bebd57891c28e8ce66b05df2877a#r66826394,
> or mute the thread
> https://github.com/notifications/unsubscribe/ABRA35IHlBUvFBMT42wKECrRwkTn-9Kfks5qLYmZgaJpZM4I0gRH
> .

afaikafk committed:
Fix bad class reference in comment and non-existent function parameter

afaikafk committed:
-Change scheduler so that it does not require leaking the task instance
out of PackagedTask's run method
-Add feature to retry failing tasks with variable retry delay and
limited number of retries

goodbadwolf commented on greentasks/scheduler.py:
```diff
@@ -1,94 +1,157 @@
+        task_instance = packaged_task.instantiate()
+        if not task_instance:
+            # instantiation failed, task is not runnable
+            return
```
We should be raising some exception here, to notify the client for the failure

goodbadwolf commented on setup.py:
```diff
@@ -1,31 +1,29 @@
-        'gevent',
-    ],
+    packages=find_packages(),
+    install_required=['gevent'],
```
Typo. Should be `install_requires`

afaikafk committed:
-Make sure periodic tasks being retried get a chance to reschedule
themselves after retry attempts are exceeded in case they are failing
-Simplify scheduler code since most of the differences in the
scheduling logic has been moved into PackagedTask
-Fix status and retry_count state changes

afaikafk committed:
Raise appropriate exception from schedule method in case task is not usable

afaikafk committed:
Fix typo in setup module
